### PR TITLE
Generate a C API for IGVM

### PR DESCRIPTION
**Pull request by roy-hopkins**
This is a proposal on how to implement https://github.com/microsoft/igvm/issues/1 and provide a C API for igvm.

The approach taken is to based on using [cbindgen](https://github.com/mozilla/cbindgen) to analyze the rust code and produce header files usable from C. To support this, a set of C friendly functions have been added to `igvm/src/c_api.rs` which allows, at present, binary files to be parsed and processed.

The additional functions are hidden behind a rust feature named `igvm-c`. When performing a normal build (without the C API) the only change that has any effect is from the first commit 4da1043900c54dd8f5a47c46df5d63d1f402453b which by necessity relaxes the unsafe_code restriction. Unfortunately this cannot be controlled by the feature.

The sample in `igvm_c/sample` and the unit test code in `igvm/tests` are probably the best places to start to see how to use the proposed API from C.

### CODE REVIEW

**Comment by chris-oo** [comment]
Thanks for the RFC! I'll try to find some time to look at this soon.

File: igvm/src/c_api.rs [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+/// # Safety
+/// This function is exported and designed to be called as a C API. As such
+/// it works with raw pointers and is inherently unsafe due to this.
+#[no_mangle]
+pub unsafe extern "C" fn igvm_new_from_binary(
+    data: *const u8,
+    len: u32,
+    igvm: *mut *mut IgvmFile,
+) -> IgvmApiError {
+    // Safety: data array must be at least as large as len
```
Comment (chris-oo) [inline igvm/src/c_api.rs:136]: nit: Can we do `SAFETY:` instead of `Safety:`. Also, we should comment exactly what the preconditions the caller must adhere to, in this case that data must be as large as len in the doc comment. Then, here in the function safety, we say "The caller guarantees that data is valid for len bytes".

Comment (chris-oo) [reply igvm/src/c_api.rs:136]: As what we really want with Safety docs here (and safety comments) is to make it clear _why_ the function is unsafe. 

And for unsafe blocks, what conditions must hold, and why this unsafe call is valid.

File: igvm/src/c_api.rs [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+/// it works with raw pointers and is inherently unsafe due to this.
+#[no_mangle]
+pub unsafe extern "C" fn igvm_get_header_type(
+    igvm: *mut IgvmFile,
+    section: IgvmHeaderSection,
+    index: u32,
+    typ: *mut IgvmVariableHeaderType,
+) -> IgvmApiError {
+    // Safety: Using pointer from C
+    unsafe {
```
Comment (chris-oo) [inline igvm/src/c_api.rs:204]: We should limit the scope of the unsafe block here as much as we can. Maybe use some local variables to create rust references, then assign the type at the very end rather than needing multiple smaller unsafe blocks throughout.

File: igvm_c/README.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+## Dependencies
+The C API header files are generated using the `cbindgen` tool. This tool needs
+to be installed before the API can be built. This can be achieved using:
+
+```bash
+cargo install --force cbindgen
+```
+
+In addition, `sample/dump_igvm` requires `gcc` to be installed.
```
Comment (chris-oo) [inline igvm_c/README.md:21]: Could we use CC instead of requiring GCC here so you could use clang/etc? Or whatever the Makefile equivalent is, not an expert here

File: igvm_c/sample/dump_igvm.c [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+static void hexdump(const void* data, size_t size, int columns, int address) {
+    int rows = (size + (columns - 1)) / columns;
+    int row;
+    int col;
+
+    for (row = 0; row < rows; ++row) {
+        printf("| %08X | ", address + row * columns);
+        for (col = 0; col < columns; ++col) {
+            size_t index = row * columns + col;
+            if (index >= size)
```
Comment (chris-oo) [inline igvm_c/sample/dump_igvm.c:26]: Even though this is sample code, I'd personally prefer braces for one line ifs.

File: igvm_c/tests/igvm_test.c [hunk 5]
```diff
@@ -0,0 +1,275 @@
+#include <CUnit/CUnit.h>
```
Comment (chris-oo) [inline igvm_c/tests/igvm_test.c:1]: Possible to wire this up in the github-actions yaml as well?

File: igvm/src/c_api.rs [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+    buf_len: *mut u32,
+) -> IgvmApiError {
+    // Safety: Using pointer from C
+    unsafe {
+        match section {
+            IgvmHeaderSection::HEADER_SECTION_PLATFORM => {}
+            IgvmHeaderSection::HEADER_SECTION_INITIALIZATION => {}
+            IgvmHeaderSection::HEADER_SECTION_DIRECTIVE => {
+                if let Some(header) = (*igvm).directive_headers.get_mut(index as usize) {
+                    let mut file_data = Vec::<u8>::new();
```
Comment (chris-oo) [inline igvm/src/c_api.rs:348]: It's somewhat unfortunate we need this _extra_ buffer. Given that we should be able to make a slice and query it without needing to do this.

Maybe we should add a method to the header types to ask for the size, so this looks cleaner?

File: igvm_c/Makefile [hunk 7]
```diff
@@ -0,0 +1,52 @@
+# SPDX-License-Identifier: MIT OR Apache-2.0
```
Comment (chris-oo) [inline igvm_c/Makefile:1]: I wonder if we could also/instead have a rust sample using the C api? Might be nicer to integrate with cargo test, as well?

Comment (chris-oo) [reply igvm_c/Makefile:1]: Or maybe just some rust tests for the C API would be good enough?

**Review by chris-oo — COMMENTED** [review]
Broad strokes look fine, but I think we can add some rust methods to remove the need to allocate Vecs that we immediately throw away for a lot of these methods.

File: igvm/src/c_api.rs (hunk 1 above)
Comment (roy-hopkins) [reply igvm/src/c_api.rs:218]: My bad. I put `Safety:` placeholders there to prevent the clippy warnings then forgot to finish the documentation before creating this PR. I've now updated the docs and switched to `SAFETY:`.

File: igvm/src/c_api.rs (hunk 2 above)
Comment (roy-hopkins) [reply igvm/src/c_api.rs]: I've managed to remove many of the unsafe regions and reworked the remainder to the smallest scope possible.

File: igvm_c/README.md (hunk 3 above)
Comment (roy-hopkins) [reply igvm_c/README.md]: Done

File: igvm_c/tests/igvm_test.c (hunk 5 above)
Comment (roy-hopkins) [reply igvm_c/tests/igvm_test.c:1]: I've added a new action that builds the C API and runs the unit tests, failing if any of the tests fail.

File: igvm/src/c_api.rs (hunk 6 above)
Comment (roy-hopkins) [reply igvm/src/c_api.rs:417]: I've reworked this significantly now. Everything is now managed by handles that hold an instance of both the main file structure and any buffers that have been returned from the query functions. This prevents unnecessary copying and also reduces the amount of unsafe code.

**Comment by roy-hopkins** [comment]
I've addressed the current review comments with the latest push. The most significant change was in the management of the buffers that are returned containing the headers and file data. Previously, the rust code was populating a `Vec` then copying the result into a C provided, pre-allocated buffer (or returning the size and `BUFFER_TOO_SMALL` if the buffer was not big enough.

I've now changed this so the `Vec` created by the rust functions are stored alongside a handle that is created when the file is initially parsed. This means that the rust code owns the buffer and knows the actual size, allowing a lot of unsafe code to be removed. Obviously to support this I have had to add functions to manage the lifetime of these handles and buffers but I think the result is safe and cleaner than the previous version.

File: igvm_c/Makefile (hunk 7 above)
Comment (chris-oo) [reply igvm_c/Makefile:1]: I'm inclined to push Rust users to the Rust API, so I think no rust sample is fine. That way no bad examples in the codebase.

File: igvm/src/c_api.rs [hunk 8]
```diff
@@ -0,0 +1,763 @@
...
+
+type IgvmResult = i64;
+type IgvmHandle = i64;
+
+struct IgvmFileInstance {
+    file: IgvmFile,
+    buffers: BTreeMap<IgvmHandle, Vec<u8>>,
+}
+
+static mut IGVM_HANDLES: Mutex<BTreeMap<IgvmHandle, IgvmFileInstance>> =
```
Comment (chris-oo) [inline igvm/src/c_api.rs:124]: I think you can remove the need to make this mutable by using std::sync::OnceLock and initializing in the igvm_create call. I would consider it an error to call the other API methods anyways since you don't have a handle. 
See: https://doc.rust-lang.org/std/sync/struct.OnceLock.html 

For example, replace it with:
```rust
static IGVM_HANDLES: OnceLock<Mutex<BTreeMap<...>
``` 

Then the corresponding usage in `igvm_create` can be modified to remove the unsafe around mutable usage to:
```rust
let handle = IGVM_HANDLE_FACTORY.get_or_init(|| AtomicI64::new(1)).fetch_add(1, Ordering::Relaxed);
let mut m = IGVM_HANDLES.get_or_init(|| Mutex::new(BTreeMap::new());
...
```

And other usages of the globals can be done with `IGVM_HANDLES.get().expect("caller called API before igvm_create")`.

Which should help us get rid of the unsafe blocks and the need for SAFETY comments.

File: igvm/src/c_api.rs [hunk 9]
```diff
@@ -0,0 +1,763 @@
...
+/// zero indicates a valid handle has been returned.
+///
+/// # Safety
+///
+/// The function assumes that there are at least `len` valid bytes of memory
+/// starting at the address pointed to by `data`. If this is violated then this
+/// will result in undefined behaviour.
+#[no_mangle]
+pub unsafe extern "C" fn igvm_new_from_binary(data: *const u8, len: u32) -> IgvmResult {
+    // Safety: data array must be at least as large as len
```
Comment (chris-oo) [inline igvm/src/c_api.rs:218]: I'd rephrase this to:
```rust
/// SAFETY: Caller guarantees that the data ptr is an array with at least len bytes of memory.
```

File: igvm/src/c_api.rs (hunk 8 above)
Comment (chris-oo) [reply igvm/src/c_api.rs:124]: Or you could return some kind of Invalid handle error also if the `.get()` fails. Maybe that's better?

File: igvm/src/c_api.rs [hunk 10]
```diff
@@ -0,0 +1,763 @@
...
+    },
+};
+
+use crate::{Error, IgvmFile};
+use open_enum::open_enum;
+
+#[open_enum]
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[repr(i64)]
+pub enum IgvmApiError {
```
Comment (chris-oo) [inline igvm/src/c_api.rs:23]: I think this is safe to return directly. Since it's `repr(transparent)` as an i64 with open_enum, there's no need for the additional IgvmResult type. That should help make the code cleaner and reduce the need for .0.

See https://doc.rust-lang.org/nomicon/other-reprs.html#reprtransparent and https://docs.rs/open-enum/0.4.0/open_enum/#integer-type.

File: igvm/src/c_api.rs [hunk 11]
```diff
@@ -0,0 +1,763 @@
...
+    index: u32,
+) -> IgvmResult {
+    let mut result = IgvmApiError::IGVMAPI_OK;
+    let mut header_binary = Vec::<u8>::new();
+
+    // SAFETY: Safe access to the IGVM_HANDLES map through a mutex.
+    let mut lock = unsafe { IGVM_HANDLES.lock().unwrap() };
+    let igvm_handle = lock.get_mut(&handle);
+
+    if let Some(igvm) = igvm_handle {
```
Comment (chris-oo) [inline igvm/src/c_api.rs:340]: I'd structure this to help reduce rightward drift in this and other functions by doing something like:
```rust
let igvm = match igvm_handle { // return the valid igvm file or early return from the function with INVALID_HANDLE };
```

That way, we avoid some rightward drift. I think also we can do that for other option usage.

However, looking at this more it seems like maybe we should refactor this so that perhaps we have rust internal functions that use Result<(), IgvmApiError>, then a wrapper conversion function to convert the Result type to the final API call. That way, we can use things like Option `ok_or(..)?` and the like to make this code less C and more Rust.

So we'd have something like:
```rust
pub extern "C" fn igvm_get_header(...) -> IgvmResult {
    match igvm_get_header_internal(...) {
        Ok(()) => IgvmResult::Ok,
        Err(e) => e, 
    }
}
``` 

Then you can write a standard rust function for `igvm_get_header_internal(...) -> Result<(), IgvmResult>`

Comment (chris-oo) [reply igvm/src/c_api.rs:340]: I think we'd find the latter to help reduce a lot of this rightward drift and `if let Some()...` usage throughout.

File: igvm/src/c_api.rs [hunk 12]
```diff
@@ -0,0 +1,763 @@
...
+            IgvmHeaderSection::HEADER_SECTION_INITIALIZATION => IgvmApiError::IGVMAPI_NO_DATA.0,
+            _ => IgvmApiError::IGVMAPI_NO_DATA.0,
+        }
+    } else {
+        IgvmApiError::IGVMAPI_INVALID_HANDLE.0
+    }
+}
+
+#[cfg(test)]
+#[allow(clippy::undocumented_unsafe_blocks)]
```
Comment (chris-oo) [inline igvm/src/c_api.rs:443]: Lets enforce SAFTEY comments on test code too, since test code is real too. Since if the test code isn't right, then we can't really trust it.

File: igvm/src/lib.rs [hunk 13]
```diff
@@ -195,6 +207,27 @@ impl IgvmPlatformHeader {
             Err(BinaryHeaderError::InvalidVariableHeaderType)
         }
     }
+
+    /// Write the binary representation of the header and any associated file
+    /// data to te supplied variable_headers and file data vectors.
```
Comment (chris-oo) [inline igvm/src/lib.rs:212]: typo: the

Comment (chris-oo) [reply igvm/src/lib.rs:212]:
```suggestion
    /// data to the supplied variable_headers and file data vectors.
```

File: igvm/src/c_api.rs [hunk 14]
```diff
@@ -0,0 +1,763 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
```
Comment (chris-oo) [inline igvm/src/c_api.rs:1]: Include a module comment with `///! ` that shows what this module is about.

File: igvm/src/lib.rs [hunk 15]
```diff
@@ -245,6 +278,21 @@ impl IgvmInitializationHeader {
         size_of::<IGVM_VHS_VARIABLE_HEADER>() + additional
     }
 
+    #[cfg(feature = "igvm-c")]
+    fn header_type(&self) -> IgvmVariableHeaderType {
```
Comment (chris-oo) [inline igvm/src/lib.rs:282]: could we add some function docs to these internal functions too?

**Comment by chris-oo** [comment]
I left another round of feedback but I agree that it's much better. I think refactoring to take advantage of `Result<(), IgvmApiError>` and `OnceLock` will help clean it up a bit more and remove all the unsafe code blocks. Thanks for doing this!

File: igvm/src/c_api.rs (hunk 8 above)
Comment (chris-oo) [reply igvm/src/c_api.rs:124]: Maybe its worth abstracting additionally the IGVM_HANDLES handout function to a rust function as well, so you can do the `get_or_init` call and also save having to copy-paste the `add(1, Ordering::Relaxed)` everywhere, too.

File: igvm/src/c_api.rs (hunk 10 above)
Comment (roy-hopkins) [reply igvm/src/c_api.rs:23]: After some thought I've now replaced the `IgvmApiError` enum with constants, resulting in `#define`'s in the generated C code. The functions now return either an `i32` or a handle type which is actually an `i32`. This makes the C return codes more obvious and still benefits from removing the additional type and the need for `.0`.

File: igvm/src/c_api.rs (hunk 8 above)
Comment (roy-hopkins) [reply igvm/src/c_api.rs:124]: I've implemented your suggestions and return an error code if `.get()` fails. This is much cleaner. Thanks.

**Comment by roy-hopkins** [comment]
I've worked through the PR comments and suggestions and uploaded a new version. In addition I've added one more commit to package and install the library.

I've been using the version built by this branch for a few weeks now working on integrating an IGVM loader into QEMU so I am happy that the C library is working well. Therefore I consider this ready for removing the RFC/draft status.

File: igvm_c/cbindgen_igvm_defs.toml [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+} IgvmPageDataFlags;
+
+typedef struct {
+	uint32_t vtl2_protectable : 1;
+	uint32_t reserved : 31;
+} RequiredMemoryFlags;
+
+typedef struct {
+	uint32_t is_shared : 1;
+	uint32_t reserved : 31;
```
Comment (roy-hopkins) [inline igvm_c/cbindgen_igvm_defs.toml:41]: This should be `uint16_t`.

File: igvm_c/sample/dump_igvm.c [hunk 17]
```diff
@@ -0,0 +1,383 @@
...
+			}
+            else {
+                char c = ((char *)data)[index];
+                if ((c >= 32) && (c < 127)) {
+                    printf("%c", c);
+				}
+                else {
+                    printf(".");
+				}
+            }
```
Comment (joergroedel) [inline igvm_c/sample/dump_igvm.c:47]: Some of the closing curly brackets are improperly aligned

**Review by joergroedel — COMMENTED** [review]
Just a minor coding style issue, otherwise this looks good to me.

**Comment by chris-oo** [comment]
This is on my radar, I'll take a look at it tomorrow. Seems like I need to fix my github email settings because I'm not getting the emails I expect.

**Review by chris-oo — COMMENTED** [review]
Thanks for cleaning up the changes! 

A few comments left, mainly mechanical, but I'm happy to sign off and spin a new release once that's done. Thanks again for contributing these changes!

File: igvm_c/sample/dump_igvm.c [hunk 18]
```diff
@@ -0,0 +1,383 @@
+/*
```
[the remark this answers is not in the document]
Comment (roy-hopkins) [reply igvm_c/sample/dump_igvm.c:1]: I've now updated all C files and the cbindgen configuration files to use spaces.

File: igvm_defs/src/lib.rs [hunk 19]
```diff
@@ -228,8 +229,6 @@ pub enum IgvmArchitecture {
 #[derive(AsBytes, FromBytes, FromZeroes, Debug, Clone, Copy, PartialEq, Eq)]
 #[repr(u32)]
 pub enum IgvmVariableHeaderType {
-    /// Invalid.
```
[the remark this answers is not in the document]
Comment (chris-oo) [reply igvm_defs/src/lib.rs:231]: My rationale is that even if it looks fairly obvious that `0` is not a free value, it's nice for us to document in the format exactly that "hey this value is invalid and don't try to use it". 

I'm fine with renaming it to IGVM_VHT_INVALID if that satisfies the C bindgen req. A bit unfortunate, though.

[the remark this answers is not in the document]
Comment (chris-oo) [reply igvm_defs/src/lib.rs:231]: I'm happy to sign off once this last little bit is in, thanks again for your changes!

**Review by joergroedel — APPROVED** [review]
Looks good to me now.

File: igvm_defs/src/lib.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (roy-hopkins) [reply igvm_defs/src/lib.rs]: I've managed to use some C preprocessor wizardry in `cbindgen_igvm_defs.toml` to remove the duplicate `INVALID` definition on the header file so I've removed the commit that deleted this `INVALID` enum entry, restoring it to how it was before.

[the remark this answers is not in the document]
Comment (chris-oo) [reply igvm_defs/src/lib.rs]: Great thanks!

**Review by chris-oo — APPROVED** [review]

**Comment by roy-hopkins** [comment]
Rebased on `microsoft:main`. No other changes.
